File: problems/cpp/deepest-leaves-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int depth = 0;
    int sum = 0;
    void go(TreeNode *node, int dep){
        if(node){
            if(dep > depth){
                depth = dep;
                sum = node->val;
            }
            else if(dep == depth){
                sum += node->val;
            }
            go(node->left, dep + 1);
            go(node->right, dep + 1);
        }
    }
    int deepestLeavesSum(TreeNode* root) {
        go(root, 0);
        return sum;
    }
};

File: problems/python/deepest-leaves-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deepestLeavesSum(self, root: Optional[TreeNode]) -> int:
        self.ans = 0
        self.depth = 0

        def go(node, d):
            if node:
                if d == self.depth:
                    self.ans += node.val
                elif d > self.depth:
                    self.depth = d
                    self.ans = node.val

                go(node.left, d + 1)
                go(node.right, d + 1)

        go(root, 0)
        return self.ans